packages/node-core/src/indexer/multiChainRewind.service.ts

```diff
@@ -0,0 +1,272 @@
+    }
+    if (rewindTimestamp) {
+      this.status = RewindStatus.Rewinding;
+      this.waitRewindHeader = await this.searchWaitRewindHeader(rewindTimestamp);
+    }
+  }
+
+  private async searchWaitRewindHeader(rewindTimestamp: number): Promise<Header> {
+    const rewindDate = dayjs(rewindTimestamp).toDate();
+    const rewindBlockHeader = await this.getHeaderByBinarySearch(rewindDate);
+    // The blockHeader.timestamp obtained from the query cannot be used directly, as it will cause an infinite loop.
+    // Different chains have timestamp discrepancies, which will result in infinite backward tracing.
+    return {...rewindBlockHeader, timestamp: rewindDate};
+  }
+
+  /**
+   * Serialize the rewind lock
+   * @param rewindTimestamp ms
+   * @param chainTotal The total number of registered chains.
+   * @returns
+   */
+  private serializeRewindLock(rewindTimestamp: number, chainTotal: number): string {
+    return JSON.stringify({timestamp: rewindTimestamp, chainNum: chainTotal});
+  }
+
+  /**
+   * If the set rewindTimestamp is greater than or equal to the current blockHeight, we do nothing because we will roll back to an earlier time.
+   * If the set rewindTimestamp is less than the current blockHeight, we should roll back to the earlier rewindTimestamp.
+   * @param rewindTimestamp rewindTimestamp in milliseconds
+   */
+  async setGlobalRewindLock(rewindTimestamp: number) {
+    const {needRewind} = await this.globalModel.setGlobalRewindLock(rewindTimestamp);
+    if (needRewind) {
+      logger.info(`setGlobalRewindLock success chainId: ${this.chainId}, rewindTimestamp: ${rewindTimestamp}`);
+      this.status = RewindStatus.Rewinding;
+    }
+  }
+
+  async releaseChainRewindLock(tx: Transaction, rewindTimestamp: number): Promise<number> {
+    const chainNum = await this.globalModel.releaseChainRewindLock(tx, rewindTimestamp);
+    // The current chain has completed the rewind, and we still need to wait for other chains to finish.
+    // When fully synchronized, set the status back to normal by pgListener.
+    this.status = RewindStatus.WaitOtherChain;
+    logger.info(`Rewind success chainId: ${JSON.stringify({chainNum, chainId: this.chainId, rewindTimestamp})}`);
+    return chainNum;
+  }
+
+  /**
+   * Get the block header closest to the given timestamp
+   * @param timestamp To find the block closest to a given timestamp
+   * @returns
+   */
+  async getHeaderByBinarySearch(timestamp: Header['timestamp']): Promise<Required<Header>> {
+    assert(timestamp, 'getHeaderByBinarySearch `timestamp` is required');
+
+    let left = 0;
+    let {height: right} = await this.storeService.getLastProcessedBlock();
+    let searchNum = 0;
+    while (left < right) {
+      searchNum++;
+      const mid = Math.floor((left + right) / 2);
+      const header = await this.blockchainService.getRequiredHeaderForHeight(mid);
+
+      if (header.timestamp === timestamp) {
+        return header;
```

Comment:
This is almost never going to be the case. Blockchains don't produce blocks with any sort of syncronisation. 

I don't think this function is doing what it should be, which is to find the nearest block header before the input timestamp